### PR TITLE
Improve Asahi Linux support

### PR DESCRIPTION
**Pull request by CarterLi**
1. ascii: display a fallback logo if the current distro logo is unavailable
2. benchmark: always print millisecond for easier comparation
3. battery: display battery model name instead of meaningless strings or indices
4. gpu: fallback to glxinfo if failed to find gpus in pci devices. Fix Asahi Linux

* Before: 
image

* After: 
image

### CODE REVIEW

File: src/ascii.rs [hunk 1]
```diff
@@ -45,6 +45,7 @@ pub fn get_ascii(os: &str) -> (String, u16) {
         "almalinux" => ALMA,
         "android" => ANDROID,
         "garuda" => GARUDA,
+        "linux" => LINUX,
```
Comment (LivacoNew) [inline src/ascii.rs:48]: Ignoring this has nothing to do with the PR, I would gate this behind a config option defaulted to off simply to draw attention to any missing ASCII's for more bug reports to add them. Once it gets reported they can get advised they use the config option as an alternative while it gets added.

File: src/ascii.rs [hunk 2]
```diff
@@ -377,6 +378,20 @@ const BAZZITE: (&str, u16) = (
           ###################        
               ###########           ", 36);
 
+const LINUX: (&str, u16) = (
```
Comment (LivacoNew) [inline src/ascii.rs:381]: What's the source for this ASCII? Or is it the same as the rest?

File: src/main.rs [hunk 3]
```diff
@@ -107,7 +107,7 @@ fn print_bench_time(benchmarking: bool, benchmark_warn: Option<u128>, name: &str
 
     // This is different to module bench times fyi
     let t: Duration = time.unwrap().elapsed();
-    let mut t_output: String = format!("{:2?}", t);
+    let mut t_output: String = format!("{:.3} ms", t.as_micros() as f64 / 1000.);
```
Comment (LivacoNew) [inline src/main.rs:110]: Not a fan of this change - would rather have more accuracy than easier units, even if it's at a scale that it hardly matters

File: src/main.rs [hunk 4]
```diff
@@ -306,18 +306,27 @@ fn main() {
...
                 print_bench_time(args.benchmark, args.benchmark_warn, "OS (for OS Color)", os_bench);
             }
-            &known_outputs.os.as_ref().unwrap().as_ref().unwrap().distro_id
+            let mut temp: CrabFetchColor = formatter::find_os_color(&known_outputs.os.as_ref().unwrap().as_ref().unwrap().distro_id);
+            if temp == CrabFetchColor::Clear {
+                temp = formatter::find_os_color(&known_outputs.os.as_ref().unwrap().as_ref().unwrap().distro_id_like)
+            }
+
+            #[cfg(target_os = "linux")]
+            if temp == CrabFetchColor::Clear {
```
Comment (LivacoNew) [inline src/main.rs:323]: You already checked this above, move this up into the same block

File: src/main.rs [hunk 5]
```diff
@@ -677,7 +686,16 @@ fn main() {
             let ascii: (String, u16) = if args.distro_override.is_some() {
                 ascii::get_ascii(&args.distro_override.clone().unwrap())
             } else {
-                ascii::get_ascii(&known_outputs.os.as_ref().unwrap().as_ref().unwrap().distro_id)
+                let mut temp: (String, u16) = ascii::get_ascii(&known_outputs.os.as_ref().unwrap().as_ref().unwrap().distro_id);
+                if temp.0.is_empty() {
+                    temp = ascii::get_ascii(&known_outputs.os.as_ref().unwrap().as_ref().unwrap().distro_id_like);
+                }
+                #[cfg(target_os = "linux")]
+                if temp.0.is_empty() {
```
Comment (LivacoNew) [inline src/main.rs:694]: Again, you already checked this

File: src/modules/battery.rs [hunk 6]
```diff
@@ -105,16 +105,16 @@ pub fn get_batteries() -> Result<Vec<BatteryInfo>, ModuleError> {
...
             Err(_) => continue,
         };
-        let id: String = match path.file_name() {
-            Some(r) => match r.to_str() {
-                Some(r) => r.strip_prefix("BAT").unwrap_or(r).to_string(),
+        let model_name: String = match util::file_read(&path.join("model_name")) {
+            Ok(r) => r.trim().to_string(),
+            Err(_) => match path.file_name() {
+                Some(r) => r.to_str().unwrap_or("Unknown").to_string(),
                 None => continue,
```
Comment (LivacoNew) [inline src/modules/battery.rs:112]: Why not return the supply name as before instead?

File: src/modules/gpu.rs [hunk 7]
```diff
@@ -110,6 +110,13 @@ pub fn get_gpus(config: &Configuration) -> Result<Vec<GPUInfo>, ModuleError> {
         Err(e) => return Err(e)
     }
 
+    if gpus.is_empty() {
+        match fill_from_glxinfo(&mut gpus) {
```
Comment (LivacoNew) [inline src/modules/gpu.rs:114]: No. This fucks performance, not doing it. Find an alternative detection.

**Review by LivacoNew — CHANGES_REQUESTED** [review]

File: src/main.rs (hunk 4 above)
Comment (CarterLi) [reply src/main.rs:323]: This is to check `temp = formatter::find_os_color(&known_outputs.os.as_ref().unwrap().as_ref().unwrap().distro_id_like)`, not `distro_id`

Comment (LivacoNew) [reply src/main.rs:323]: `if temp.0.is_empty() {` gets checked in both scenarios - Move `temp = formatter::find_os_color("linux")` up into the same block with your conditional compile so you only ever check it once - use the `cfg!` macro if you need

File: src/ascii.rs (hunk 2 above)
Comment (CarterLi) [reply src/ascii.rs:381]: Copied from neofetch.

File: src/main.rs (hunk 3 above)
Comment (CarterLi) [reply src/main.rs:110]: This is for comparation. You can find the bottleneck at the first glance if the all have the same unit.

File: src/modules/battery.rs (hunk 6 above)
Comment (CarterLi) [reply src/modules/battery.rs:112]: This is the fallback if we fail to acquire the model name. A string is better than just an index.

In addition, it's `macsmc-battery` for asahi, not `BATx`

File: src/modules/gpu.rs (hunk 7 above)
Comment (CarterLi) [reply src/modules/gpu.rs:114]: Fuck what? It fucks only if `fill_from_pcisysfile` fails, so it only fucks ARM devices.

**Comment by CarterLi** [comment]
Please don't change the title. If you hate this PR. just close it.

File: src/main.rs (hunk 4 above)
Comment (CarterLi) [reply src/main.rs:323]: `temp = formatter::find_os_color(&known_outputs.os.as_ref().unwrap().as_ref().unwrap().distro_id_like)` Can also fail. How can I check it if I remove the `if` check?

File: src/modules/battery.rs (hunk 6 above)
Comment (LivacoNew) [reply src/modules/battery.rs:112]: Ah nvm I missed it somehow, ignore me lol

File: src/modules/gpu.rs (hunk 7 above)
Comment (LivacoNew) [reply src/modules/gpu.rs:114]: Yes, but I refuse to believe there's not better ways of detecting the name from arm than using glxinfo.

File: src/main.rs (hunk 3 above)
Comment (LivacoNew) [reply src/main.rs:110]: Add it as a command line argument and let the user decide then, don't force it though

File: src/ascii.rs (hunk 2 above)
Comment (LivacoNew) [reply src/ascii.rs:381]: Add a comment then please to source it

**Comment by LivacoNew** [comment]
> Please don't change the title. If you hate this PR. just close it.

I changed it because prefixing ***[GIFT]*** is just unnecessary, not because I think the PR is shit. Don't get sensitive over a title lol

**Pull request closed without merging**